backend/analyze_tailcall_graph.ml

```diff
@@ -0,0 +1,416 @@
+    type elt = T.key
+    type t = unit T.t
+
+    let create n = T.create n
+    let add t k = T.replace t k ()
+    let remove t k = T.remove t k
+    let mem t k = T.mem t k
+    let iter t ~f = T.iter (fun e () -> f e) t
+  end
+end
+
+module Graph : sig
+  module Vertex : sig
+    type t
+
+    val id : t -> int
+    val unknown : t
+    val is_unknown : t -> bool
+    val to_dot_id : t -> string
+    val to_dot_label : t -> string
+
+    module Tbl : Hashtbl.S with type key = t
+    module Hashset : Hashset.S with type elt = t
+  end
+
+  module Edge : sig
+    (* For the purpose of analyzing whether TCO inference might break existing
+       code (by causing a stack overflow) we are interested in whether there
+       are any cycles with Plain_tail_edges and Inferred_nontail_edges.
+       To find these cycles, we plan on finding SCC's for the subgraph
+       consisting of just edges with those labels. *)
+    type label =
+      (* "Plain" edges are edges that are (very likely) not changing their
+         TCO behavior as a result of adding TCO inference. I.e., before TCO
+         inference, they were tail (nontail) if and only if after TCO
+         inference they are tail (nontail). *)
+      | Plain_tail_edge
+      | Plain_nontail_edge
+      (* "Inferred" edges are edges that might possibly change their TCO
+         behavior as a result of adding TCO inference. *)
+      | Inferred_tail_edge
+      | Inferred_nontail_edge
+
+    type t =
+      { from : Vertex.t
+      ; to_ : Vertex.t
+      ; label : label
+      }
+
+    module Tbl : Hashtbl.S with type key = t
+    module Hashset : Hashset.S with type elt = t
+  end
+
+  type t
+
+  val create : unit -> t
+  val reset : t -> unit
+  val successors : t -> Vertex.t -> Edge.Hashset.t
+  val find_or_add_vertex : t -> fn_name:string -> Vertex.t
+
+  val add_edge
+    :  t
+    -> from:Vertex.t
+    -> to_:Vertex.t
+    -> actual_position:[ `Tail | `Nontail ]
+    -> original_position:Typedtree.position_and_tail_attribute
+    -> unit
+
+  val print_dot : t -> Format.formatter -> unit
+end = struct
+  module Vertex = struct
+    module T = struct
+      type t =
+        | Unknown_fn
+        | Known_fn of
+            { id : int
+            ; name : string
+            }
+
+      let unknown = Unknown_fn
+
+      let is_unknown t =
+        match t with
+        | Unknown_fn -> true
+        | Known_fn _ -> false
+      ;;
+
+      let id t =
+        match t with
+        | Unknown_fn -> -1
+        | Known_fn { id; _ } -> id
+      ;;
+
+      let equal t1 t2 = id t1 = id t2
+      let hash t = Int.hash (id t)
+    end
+
+    include T
+
+    let to_dot_id t =
+      match t with
+      | Unknown_fn -> "unknown"
+      | Known_fn { id; _ } -> Int.to_string id
+    ;;
+
+    let to_dot_label t =
+      match t with
+      | Unknown_fn -> "<unknown>"
+      | Known_fn { name; _ } -> name
+    ;;
+
+    module Tbl = Hashtbl.Make (struct
+        include T
+      end)
```

Comment:
Why not `Hashtbl.Make (T)`? Same below.

Reply:
Because of refactoring (previously I only defined hash in this anonymous struct, and then I pulled it out.)